pkg/controller/syncmachineset/syncmachineset_controller.go

```diff
@@ -0,0 +1,555 @@
+	// secret. Until then we'll look for a
+	// "clustername-kubeconfig" secret.
+	//
+	// clusterConfig := cluster.Status.AdminKubeconfig
+	// if clusterConfig == nil {
+	// 	return nil, fmt.Errorf("unable to retrieve AdminKubeConfig from cluster status")
+	// }
+	secretName := cluster.Name + "-kubeconfig"
+
+	// Step 3: Retrieve secret
+	secret, err := c.kubeClient.CoreV1().Secrets(machineSet.Namespace).Get(secretName, metav1.GetOptions{})
+	if err != nil {
+		return nil, nil, err
+	}
+	secretData := secret.Data["kubeconfig"]
+
+	// Step 4: Generate config from secret data
+	config, err := clientcmd.Load(secretData)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	kubeConfig := clientcmd.NewDefaultClientConfig(*config, &clientcmd.ConfigOverrides{})
+	restConfig, err := kubeConfig.ClientConfig()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Step 5: Return client for remote cluster
+	clusterAPIClient, err := clusterapiclient.NewForConfig(restConfig)
+	if err != nil {
+		return nil, nil, err
+	}
+	deploymentClient, err := kubeclientset.NewForConfig(restConfig)
+	if err != nil {
+		return nil, nil, err
+	}
+	return clusterAPIClient, deploymentClient, nil
+}
+
+// syncMachineSet will sync the machine set with the given key.
+// This function is not meant to be invoked concurrently with the same key.
+func (c *Controller) syncMachineSet(queuedMS queuedMachineSet) error {
+	startTime := time.Now()
+	key := queuedMS.key
+	c.logger.WithField("key", key).Debug("syncing machineset")
+	defer func() {
+		c.logger.WithFields(log.Fields{
+			"key":      key,
+			"duration": time.Now().Sub(startTime),
+		}).Debug("finished syncing machineset")
+	}()
+
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+	machineSet, err := c.machineSetsLister.MachineSets(namespace).Get(name)
+	if errors.IsNotFound(err) {
+		c.logger.WithField("key", key).Debug("machineset has been deleted")
+		// TODO: ensure deleted in remote cluster. If so, how
+		// are we going to figure out what cluster the machine
+		// set was from to load the correct kubeconfig secret?
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	// Master machinesets are managed by the CO cluster and are
+	// not synced with the remote cluster. When syncing a master
+	// machineset, we will sync the cluster that the master
+	// machineset belongs to rather than the master machineset
+	// itself.
+	if isMasterMachineSet(machineSet) {
+		return c.syncMasterMachineSet(machineSet)
+	}
+
+	return c.syncComputeMachineSet(machineSet)
+}
+
+// Creates the cluster in the remote.
+func (c *Controller) syncMasterMachineSet(ms *cov1.MachineSet) error {
+	msLog := logging.WithMachineSet(c.logger, ms)
+	msLog.Debug("loaded master machineset")
+
+	// handleErr will requeue machine sets after waiting when the
+	// remote cluster API is not installed.
+	if !ms.Status.ClusterAPIInstalled {
+		return NewErrClusterAPINotInstalled(errorMsgClusterAPINotInstalled)
```

Comment:
machine sets take a long time to sync initially since they get dropped while waiting for the remote cluster API to be ready so this was a workaround to make them sync as soon as the remote API is good. Based on https://github.com/openshift/cluster-operator/pull/183#discussion_r187442333, it sounds like the alternative would be to enqueue compute sets gathered from the cluster object when we're dealing with the master machine set rather than separately?

Reply:
I don't quite understand this comment. My point with my comment was that the machine set will already be requeued by the informer when the `ClusterAPIInstalled` field is changed. There is no need to requeue it on a timer. In other words, if the controller sees that `ClusterAPIInstalled` is false and the controller has not been notified by the informer that the machine set has changed, then the `ClusterAPIInstalled` field will still be false on the check 20 seconds later.

Reply:
Ahhh, I'll remove the requeue from master sets. `ClusterAPIInstalled` isn't currently set on non master machine sets so they don't get requeued. I'll need to do something else to avoid requeueing on wait with compute sets.

Reply:
For non-master machine sets, they should be requeued when the master machine set for the cluster is updated. In `addMachineSet` and `updateMachineSet` if the machine set being added/updated is a master machine set, then the other non-master machine sets for the cluster should be enqueued.